### PR TITLE
feat(new): FormField component

### PR DESCRIPTION
**Pull request by ryancarville**
New `FormField` component that serves as a `parent` to any `input` element.  Key props are `label`, `status`, `helperText`, `error` & `errorText`.

ClickUp Card
https://app.clickup.com/t/8692u2jz7

### CODE REVIEW

File: src/components/FormField/FormField.tsx [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    errorTextColor = Validation.Error,
+    helperText,
+    helperTextColor = Validation.Default,
+    hiddenLabel = false,
+    label,
+    readOnly = false,
+    status = Validation.Default,
+    'data-test-id': dataTestId = 'fondue-form-field',
+}: FormFieldProps): ReactElement => {
+    const getFormattedLabel = (): ReactElement => {
```
Comment (bojangles-m) [inline src/components/FormField/FormField.tsx:50]:
```suggestion
    const FormattedLabel = (): ReactElement => {
```

File: src/components/FormField/FormField.tsx [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                'tw-flex tw-text-text-weak tw-gap-2 tw-select-none tw-max-w-full',
+                label.hugWidth ? 'tw-w-fit tw-items-start tw-justify-start' : 'tw-w-full tw-flex-col',
+                disabled || readOnly
+                    ? 'hover:tw-cursor-not-allowed tw-pointer-events-none'
+                    : 'hover:tw-cursor-pointer hover:tw-text-text group-hover:tw-text-text',
+            ])}
+            data-test-id={`${dataTestId}-label`}
+            aria-label={label.text}
+        >
+            {hiddenLabel ? null : getFormattedLabel()}
```
Comment (bojangles-m) [inline src/components/FormField/FormField.tsx:112]:
```suggestion
            {hiddenLabel ? null : <FormattedLabel />}
```

File: src/components/FormField/FormField.tsx [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+                </span>,
+            );
+        } else {
+            elements.push(<span>{formattedLabel}</span>);
+        }
+
+        if (applySecondaryLabel) {
+            elements.push(<span>{secondaryLabel}</span>);
+        }
+        return (
```
Comment (bojangles-m) [inline src/components/FormField/FormField.tsx:84]:
```suggestion

        return (
```

File: src/components/FormField/FormField.tsx [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        const elements: ReactElement[] = [];
+        const secondaryLabelStyle = 'tw-w-full tw-flex tw-flex-nowrap tw-justify-between';
+        const applySecondaryLabel = !hugWidth && secondaryLabel;
+        let formattedLabel = text;
+
+        if (required) {
+            formattedLabel += ' *';
+        }
+        if (tooltips?.length) {
+            const tooltipElements: ReactElement[] = tooltips.map((tip: TooltipProps) => (
```
Comment (bojangles-m) [inline src/components/FormField/FormField.tsx:61]: You can make this map an function.

```suggestion
            const FormatTooltip = () =>
            tooltips?.map((tip: TooltipProps) => (
                <Tooltip
                    key={`form-field-tooltip-${generateRandomId()}`}
                    content={tip.content}
                    size={tip.size}
                    data-test-id={tip['data-test-id']}
                >
                    {tip.children}
                </Tooltip>
            ));
```

Outside of the if statement.

File: src/components/FormField/FormField.tsx [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+                    size={tip.size}
+                    data-test-id={tip['data-test-id']}
+                >
+                    {tip.children}
+                </Tooltip>
+            ));
+            elements.push(
+                <span className={'tw-flex tw-gap-2'}>
+                    {formattedLabel}
+                    {...tooltipElements}
```
Comment (bojangles-m) [inline src/components/FormField/FormField.tsx:74]:
```suggestion
                   <FormatTooltip />
```
And then here call it.

**Review by bojangles-m — COMMENTED** [review]
Great job, just a small suggestion I posted.

**Comment by ryancarville** [comment]
> Great job, just a small suggestion I posted.

Thank you @bojangles-m good suggestion. Will make sure to address them.

File: src/components/FormField/FormField.tsx (hunk 2 above)
Comment (ryancarville) [reply src/components/FormField/FormField.tsx:112]: @bojangles-m the issue I see with abstracting this method to a component is that is it very `prop` dependant and would be a bit "messy" to move it outside the main render.

Comment (bojangles-m) [reply src/components/FormField/FormField.tsx]: YOu dont need to extract it, just change the name :)

File: src/components/FormField/FormField.tsx [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    hiddenLabel?: boolean;
+    readOnly?: boolean;
+    status?: Validation;
+    'data-test-id'?: string;
+};
+
+type FormatTooltipsProps = { tooltips: TooltipProps[] };
+
+const FormatTooltips = ({ tooltips }: FormatTooltipsProps): ReactElement[] => {
+    return tooltips.map((tip: TooltipProps) => (
```
Comment (bojangles-m) [inline src/components/FormField/FormField.tsx:39]:
```suggestion
tooltips.map((tip: TooltipProps) => (
```
remove the return, you have only one child

File: src/components/FormField/FormField.tsx (hunk 2 above)
Comment (ryancarville) [reply src/components/FormField/FormField.tsx]: SonarLint does not like this approach (and i agree).  
```Do not define components during render. React will see a new component type on every render and destroy the entire subtree\u8217s DOM nodes and state. Instead, move this component definition out of the parent component \u8220FormField\u8221 and pass data as props.```

File: src/components/FormField/FormField.stories.tsx [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    const currentValidation = args.error ? Validation.Error : Validation.Success;
+    return (
+        <FormField
+            {...args}
+            label={{
+                text: 'Number Greater than 0',
+                secondaryLabel: '1/10',
+                required: true,
+            }}
+            // error={error}
```
Comment (bojangles-m) [inline src/components/FormField/FormField.stories.tsx:95]: If it is not needed remove the comment, please

**Review by bojangles-m — APPROVED** [review]

**Review by bojangles-m — APPROVED** [review]

File: src/components/FormField/FormField.tsx (hunk 2 above)
Comment (ryancarville) [reply src/components/FormField/FormField.tsx]: Made the change to a component. 👍

Comment (bojangles-m) [reply src/components/FormField/FormField.tsx]: great

**Pull request merged**